src/dune_engine/scheduler.ml

```diff
@@ -677,21 +694,45 @@ module Run_once : sig
 end = struct
   type run_error =
     | Got_signal
-    | Files_changed
     | Never
     | Exn of Exn_with_backtrace.t
 
   exception Abort of run_error
 
-  let iter () =
-    let count = Event.pending_jobs () in
-    if count = 0 then
+  (** This function is the heart of the scheduler. It makes progress in
+      executing fibers by doing the following:
+
+      - notifying completed jobs
+      - starting cancellations
+      - terminating the scheduler on signals *)
+  let rec iter (t : t) =
+    (* In polling mode, we always wait until the next job starts, so there's no
+       condition for termination *)
+    if (not t.polling) && Event.pending_jobs () = 0 then
       raise (Abort Never)
     else (
       Console.Status_line.refresh ();
       match Event.next () with
       | Job_completed (job, status) -> Fiber.Fill (job.ivar, status)
-      | Files_changed -> raise (Abort Files_changed)
+      | Files_changed -> (
+        match t.status with
+        | Cancelling_build ->
+          (* We're already cancelling build, so file change events don't matter *)
+          iter t
+        | Building ->
+          let status_line =
+            Some
+              (Pp.seq
+                 (* XXX Why do we print "Had errors"? The user simply edited a
+                    file *)
+                 (Pp.tag User_message.Style.Error (Pp.verbatim "Had errors"))
+                 (Pp.verbatim ", killing current build..."))
+          in
+          Console.Status_line.set (Fun.const status_line);
+          t.status <- Cancelling_build;
+          Process_watcher.killall Sys.sigkill;
```

Comment:
I just spent some time trying to prove that it's not possible to spawn a process after `killall` happened and have that process "escape" the check.

I think that's indeed impossible(-ish?) because processes get spawned (in `Process.run_internal`) immediately after `Scheduler.with_job_slot` gives you a slot, without anything intervening (and we don't have multiple processes sharing a slot).

I wish it was more obvious, but I don't have concrete ideas.